### DISCUSSION TITLE
OpenTK control for Eto.Forms?

### DISCUSSION THREAD

**Discussion by philstopford** [opentk/opentk#1691] [Q&A]
Way back, there was an effort to provide OpenTK inside Eto.Forms, but this ran into some trouble following refactorings in OpenTK (https://github.com/picoe/Eto.OpenTK). I wondered whether there was a way to use recent versions of OpenTK in a similar manner, but can't see the way forward.

**Comment by NogginBops** [comment]
What issues exist with `Eto.OpenTK`? I have zero knowledge about how Eto works or what the issues is for getting it working.
What kind of refactoring make `Eto.OpenTK` hard to get working?

**Comment by philstopford** [comment]
Hello @NogginBops : this is the previous discussion (https://github.com/picoe/Eto.OpenTK/issues/40) that describes some of the issues previously run into and Curtis and Rob were bringing up various questions. I went to look back at it and I think you participated at the time. From a review, I think the questions look like this:

- What might it take to get OpenTK to work within WinForms/WPF?
- What might it take to get OpenTK working under macOS?
- What might it take to get OpenTK to work within GTK3? The last effort here was with GTK2 and I suspect a lot has changed.
- How is line drawing performed now (seemed to need some form of shader rather than what was available before OpenTK4)?

I'd like to be able to make some meaningful contribution here, but I've no idea where to start.

**Reply by NogginBops** [reply]
There is a WinForms component in the form of https://github.com/opentk/GLControl and there is a WPF component in the form of https://github.com/opentk/GLWpfControl.
macOS should just work if you set the requested OpenGL version to 4.1 and set the ForwardCompatible flag in the `NativeWindowSettings`.
I have no idea how complicated it is to get GTK3 running with OpenGL, if GTK3 can give you an opengl context then it is as simple as implementing `IBindingsContext` but if it doesn't it might be more complicated.

**Reply by NogginBops** [reply]
For line drawing you would in modern opengl indeed use a shader in combination with `glDrawArrays`/`glDrawElements` with the appropriate `GL_LINE` arguments to tell OpenGL that you want to render lines.

**Reply by philstopford** [reply]
I am definitely missing something. I tried putting one of the learning exercise into the Eto.OpenTK example code and it fails because bindings are not initialized (https://github.com/philstopford/Eto.OpenTK/tree/init_fail), but only when shader code is involved (it doesn't fail with prior GL calls). I'm not clear on what needs to be done.

**Comment by philstopford** [comment]
I actually noticed a long-open PR here : https://github.com/picoe/Eto.OpenTK/pull/41

That seems to work in GTK3. It remains unclear to me how lines are to be drawn.
